### PR TITLE
Add an dataset based on 'deep image descriptors'.

### LINKED ISSUE #143: Consider adding a 'deep image descriptor' dataset
In the past few years, it seems that the computer vision community transitioned away from representing images using traditional features like SIFT. It's now common to create 'deep descriptors' of images through convolutional neural nets (for example models based on the Inception architecture). It would be helpful to include a dataset of deep image descriptors to more closely represent this approach to image similarity search.

One idea is to take a sample from [this large dataset](http://sites.skoltech.ru/compvision/noimi/), which consists of images from ImageNet run through the a CNN to produce vectors. Some prominent kNN papers report results against this dataset, including [Billion-scale similarity search with GPUs](https://arxiv.org/abs/1702.08734) and [Efficient and robust approximate nearest neighbor search using Hierarchical Navigable Small World graphs](https://arxiv.org/abs/1603.09320).

### PR DESCRIPTION
**Pull request by jtibshirani**
This commit adds an image similarity dataset based on the data from 'Efficient
Indexing of Billion-Scale datasets of deep descriptors'. The data is composed of
images from the web that have been run through a 'deep descriptor network' to
produce vectors. The vectors were then reduced to 96 dimensions through PCA and
finally L2-normalized. We use a 10 million vector sample of the full dataset.

The original dataset can be downloaded from
http://sites.skoltech.ru/compvision/noimi/. The 10 million vectors are taken from
the file deep10M.fvecs. I emailed the authors of the dataset (with @erikbern in CC)
and they're okay with us committing a sample of the dataset.

Closes #143.

### CODE REVIEW

File: ann_benchmarks/datasets.py [hunk 1]
```diff
@@ -192,6 +194,33 @@ def fashion_mnist(out_fn):
     test = _load_mnist_vectors('fashion-mnist-test.gz')
     write_output(train, test, out_fn, 'euclidean')
 
+# Creates a 'deep image descriptor' dataset using the 'deep10M.fvecs' sample
+# from http://sites.skoltech.ru/compvision/noimi/. The download logic is adapted
+# from the script https://github.com/arbabenko/GNOIMI/blob/master/downloadDeep1B.py.
+def deep_image(out_fn):
+    train_size = 10 * 1000 * 1000
+    test_size = 1000
```
Comment (erikbern) [inline ann_benchmarks/datasets.py]: could we make the test set a bit bigger? 1k items make the recall numbers less reliable

Comment (erikbern) [reply ann_benchmarks/datasets.py]: according to README.md this should be 10k also :)

File: ann_benchmarks/datasets.py [hunk 2]
```diff
@@ -192,6 +194,33 @@ def fashion_mnist(out_fn):
...
+    # In the fvecs file format, each vector is stored by first writing its
+    # length as an integer, then writing its components as floats.
+    fv = numpy.fromfile(filename, dtype=numpy.float32)
+    dim = fv.view(numpy.int32)[0]
+
+    limit = (dim + 1) * (train_size + test_size)
+    fv = fv[0:limit]
+    X_train = fv.reshape(-1, dim + 1)[:, 1:]
+
+    X_train, X_test = train_test_split(X, test_size=test_size)
```
Comment (erikbern) [inline ann_benchmarks/datasets.py]: the code currently fails on this line since `X` isn't defined. should L220 say `X` rather than `X_train`?

**Comment by jtibshirani** [comment]
I'm sorry @erikbern for being sloppy here, I accidentally pushed a half-reverted refactor. The test set now has the correct size, and the dataset creation should run without error.

**Comment by erikbern** [comment]
will take a look shortly. i tried to run this on my computer but it crashed because it was taking so much RAM :) might have to rent an EC2 instance

**Comment by jtibshirani** [comment]
Great, thanks! I wanted to note there's a smaller sample file available called `deep10M.vecs` with a total of 10M vectors. It's ~3.6GB compared to the ~10GB file we're currently downloading. We could easily switch to using `deep10M.vecs` if it makes your workflow easier.

**Comment by erikbern** [comment]
> Great, thanks! I wanted to note there's a smaller sample file available called `deep10M.vecs` with a total of 10M vectors. It's ~3.6GB compared to the ~10GB file we're currently downloading. We could easily switch to using `deep10M.vecs` if it makes your workflow easier.

If it's easy then I'd prefer that – I think more importantly it makes the choice of 10M vectors a bit less "arbitrary". But the faster download speed doesn't hurt!

**Comment by jtibshirani** [comment]
I just updated to `deep10M.fvecs`. I initially thought it'd be nice to use the `base` files since it would easily allow for creating larger datasets in the future, but that might've been thinking too far ahead :)

**Comment by erikbern** [comment]
this is great. i'll try to merge this weekend!

**Comment by erikbern** [comment]
I was able to run this code and generate a dataset that I uploaded to http://ann-benchmarks.com/deep-image-96-angular.hdf5

Let's merge!
